projects/extension/src/background/ConnectionManager.ts

```diff
@@ -229,78 +229,143 @@ export class ConnectionManager extends (EventEmitter as { new(): StateEmitter })
+      n.name.toLowerCase() === name.toLowerCase() && n.tabId === tabId
+    );
     const network: Network = {
       tabId: tabId || 0,
-      name,
+      id,
+      name: name.toLowerCase(),
       chain: addedChain,
       status: 'connected'
     }
+    !existingNetwork && this.#networks.push(network);
+    return existingNetwork || network;
+  }
 
-    this.#networks.push(network);
+  #initHealthChecker = (app: App, isParachain?: boolean): void => {
+    if (isParachain) {
+      // eslint-disable-next-line @typescript-eslint/unbound-method
+      app.parachain && app.healthChecker?.setSendJsonRpc(app.parachain.sendJsonRpc);
+    } else {
+      // eslint-disable-next-line @typescript-eslint/unbound-method
+      app.chain && app.healthChecker?.setSendJsonRpc(app.chain.sendJsonRpc);
+    }
+    void app.healthChecker?.start((health: SmoldotHealth) => {
+      if (health && (
+        app.healthStatus?.peers !== health.peers ||
+        app.healthStatus.isSyncing !== health.isSyncing
+        )) {
+          this.emit('appsChanged', this.apps);
+      }
+      app.healthStatus = health
+    });
+    // process any RPC requests that came in while waiting for `addChain` to complete
+    if (app.pendingRequests.length > 0) {
+      app.pendingRequests.forEach(req => app.healthChecker?.sendJsonRpc(req));
+      app.pendingRequests = [];
+    }
+  }
 
-    return network;
+  #handleError = (app: App, e: Error): void => {
+    const error: MessageFromManager = { type: 'error', payload: e.message };
+    app.port.postMessage(error);
+    app.port.disconnect();
+    this.unregisterApp(app);
   }
 
   /** Handles the incoming message that contains Spec. */
   #handleSpecMessage = (msg: MessageToManager, app: App): void => {
-    const chainSpec: string = relayChains.has(app.chainName) ?
-      (relayChains.get(app.chainName) || '') : msg.payload;
+    const chainSpec: string = relayChains.has(app.chainName.toLowerCase()) ?
+      (relayChains.get(app.chainName.toLowerCase()) || '') : msg.payload;
 
     const rpcCallback = (rpc: string) => {
-      const rpcResp = app.healthChecker?.responsePassThrough(rpc);
+      const rpcResp: string | null | undefined = app.healthChecker?.responsePassThrough(rpc);
       if (rpcResp)
         app.port.postMessage({ type: 'rpc', payload: rpcResp })
     }
+    
+    let chainPromise: Promise<void>;
+
+    // Means this is a parachain trying to connect
+    if (msg.parachainPayload) {
+      // Connect the main Chain first and on success the parachain with the chain
+      // that just got connected as the relayChain
+      const relayId: string = JSON.parse(msg.parachainPayload).relay_chain;
+      const parachainSpec: string = msg.parachainPayload;
+
+      const relaychainSpec: string | undefined = relayChains.get(
+        nameIdMapper.get(relayId) || ""
+      );
+
+      if (!relaychainSpec) {
+        const error: Error = new Error("Relay chain spec was not found");
+        this.#handleError(app, error);
```

Comment:
You need to return here to stop further processing of the message

Reply:
This ought to have a test. Send a spec message with a chainspec containing an invalid relay chain and check the port gets sent an error and disconnected